File: src/Solana/Entry.cpp
// Copyright © 2017-2020 Trust Wallet.
//
// This file is part of Trust. The full Trust copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.

#include "Entry.h"

#include "Address.h"
#include "Signer.h"

using namespace TW;
using namespace std;

namespace TW::Solana {

// Note: avoid business logic from here, rather just call into classes like Address, Signer, etc.

bool Entry::validateAddress([[maybe_unused]] TWCoinType coin, const string& address, TW::byte, TW::byte, const char*) const {
    return Address::isValid(address);
}

string Entry::deriveAddress([[maybe_unused]] TWCoinType coin, const PublicKey& publicKey, TW::byte, const char*) const {
    return Address(publicKey).string();
}

Data Entry::addressToData([[maybe_unused]] TWCoinType coin, const std::string& address) const {
    return Address(address).vector();
}

void Entry::sign([[maybe_unused]] TWCoinType coin, const TW::Data& dataIn, TW::Data& dataOut) const {
    signTemplate<Signer, Proto::SigningInput>(dataIn, dataOut);
}

string Entry::signJSON([[maybe_unused]] TWCoinType coin, const std::string& json, const Data& key) const {
    return Signer::signJSON(json, key);
}

TW::Data Entry::preImageHashes([[maybe_unused]] TWCoinType coin, const TW::Data& txInputData) const {
    return txCompilerTemplate<Proto::SigningInput, Proto::PreSigningOutput>(
        txInputData, [](const auto& input, auto& output) {
            auto signer = Signer(input);
            auto preimageHash = signer.preImageHash();
            // for Solana, there is no need to hash data.
            output.set_data(preimageHash.data(), preimageHash.size());
            auto signers = signer.signers();
            auto nSigners = output.mutable_signers();
            for (auto i = 0ul; i < signers.size();i++) {
                auto newSigner = nSigners->Add();
                *newSigner = signers[i];
            }
        });
}

void Entry::compile([[maybe_unused]] TWCoinType coin, const Data& txInputData, const std::vector<Data>& signatures,
                    const std::vector<PublicKey>& publicKeys, Data& dataOut) const {
    dataOut = txCompilerTemplate<Proto::SigningInput, Proto::SigningOutput>(
        txInputData, [&](const auto& input, auto& output) {
            auto signer = Signer(input);
            output = signer.compile(signatures, publicKeys);
        });
}

} // namespace TW::Solana

File: rust/tw_internet_computer/src/entry.rs
// Copyright © 2017-2023 Trust Wallet.
//
// This file is part of Trust. The full Trust copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.

use tw_coin_entry::{
    coin_context::CoinContext,
    coin_entry::CoinEntry,
    error::{AddressError, AddressResult},
    modules::{json_signer::NoJsonSigner, plan_builder::NoPlanBuilder},
    prefix::NoPrefix,
};

use tw_proto::InternetComputer::Proto;
use tw_proto::TxCompiler::Proto as CompilerProto;

use crate::{
    context::StandardInternetComputerContext, icp::address::AccountIdentifier,
    principal::Principal, signer::Signer,
};

pub struct InternetComputerEntry;

impl CoinEntry for InternetComputerEntry {
    type AddressPrefix = NoPrefix;

    type Address = AccountIdentifier;

    type SigningInput<'a> = Proto::SigningInput<'a>;

    type SigningOutput = Proto::SigningOutput<'static>;

    type PreSigningOutput = CompilerProto::PreSigningOutput<'static>;

    type JsonSigner = NoJsonSigner;

    type PlanBuilder = NoPlanBuilder;

    #[inline]
    fn parse_address(
        &self,
        _coin: &dyn CoinContext,
        address: &str,
        _prefix: Option<Self::AddressPrefix>,
    ) -> AddressResult<Self::Address> {
        AccountIdentifier::from_hex(address).map_err(|_| AddressError::FromHexError)
    }

    #[inline]
    fn derive_address(
        &self,
        _coin: &dyn tw_coin_entry::coin_context::CoinContext,
        public_key: tw_keypair::tw::PublicKey,
        _derivation: tw_coin_entry::derivation::Derivation,
        _prefix: Option<Self::AddressPrefix>,
    ) -> tw_coin_entry::error::AddressResult<Self::Address> {
        let secp256k1_public_key = public_key
            .to_secp256k1()
            .ok_or(AddressError::PublicKeyTypeMismatch)?;
        let principal = Principal::from(secp256k1_public_key);
        let address = AccountIdentifier::new(&principal);
        Ok(address)
    }

    #[inline]
    fn sign(
        &self,
        _coin: &dyn tw_coin_entry::coin_context::CoinContext,
        input: Self::SigningInput<'_>,
    ) -> Self::SigningOutput {
        Signer::<StandardInternetComputerContext>::sign_proto(input)
    }

    fn preimage_hashes(
        &self,
        _coin: &dyn tw_coin_entry::coin_context::CoinContext,
        _input: Self::SigningInput<'_>,
    ) -> Self::PreSigningOutput {
        unimplemented!()
    }

    fn compile(
        &self,
        _coin: &dyn tw_coin_entry::coin_context::CoinContext,
        _input: Self::SigningInput<'_>,
        _signatures: Vec<tw_coin_entry::coin_entry::SignatureBytes>,
        _public_keys: Vec<tw_coin_entry::coin_entry::PublicKeyBytes>,
    ) -> Self::SigningOutput {
        unimplemented!()
    }
}
